what4/src/What4/Serialize/SETokens.hs

```diff
@@ -0,0 +1,265 @@
+import qualified Data.SCargot.Comments as SC
+import qualified Data.SCargot.Repr as SC
+import qualified Data.SCargot.Repr.WellFormed as SC
+import           Data.Semigroup
+import qualified Data.Sequence as Seq
+import           Data.Text (Text)
+import qualified Data.Text as T
+import qualified LibBF as BF
+import           Numeric.Natural ( Natural )
+import qualified Text.Parsec as P
+import           Text.Parsec.Text ( Parser )
+import           Text.Printf ( printf )
+import           Data.Ratio
+
+import           Data.Parameterized.Some ( Some(..))
+import qualified What4.BaseTypes as W4
+
+import           Prelude
+
+data Atom =
+  AId Text
+  -- ^ An identifier.
+  | AStr (Some W4.StringInfoRepr) Text
+  -- ^ A prefix followed by a string literal
+  -- (.e.g, AStr "u" "Hello World" is serialize as `#u"Hello World"`).
+  | AInt Integer
+  -- ^ Integer (i.e., unbounded) literal.
+  | ANat Natural
+  -- ^ Natural (i.e., unbounded) literal
+  | AReal Rational
+  -- ^ Real (i.e., unbounded) literal.
+  | AFloat (Some W4.FloatPrecisionRepr) BF.BigFloat
+  -- ^ A floating point literal (with precision)
+  | ABV Int Integer
+  -- ^ Bitvector, width and then value.
+  | ABool Bool
+  -- ^ Boolean literal.
+  deriving (Show, Eq, Ord)
+
+type SExpr = SC.WellFormedSExpr Atom
+
+string :: Some W4.StringInfoRepr -> Text -> SExpr
+string strInfo str = SC.A $ AStr strInfo str
+
+string' :: Some W4.StringInfoRepr -> String -> SExpr
+string' strInfo str = SC.A $ AStr strInfo (T.pack str)
+
+-- | Lift an unquoted identifier.
+ident :: Text -> SExpr
+ident = SC.A . AId
+
+ident' :: String -> SExpr
+ident' = SC.A . AId . T.pack
+
+
+-- | Lift a quoted identifier.
+-- quoted :: String -> SExpr
+-- quoted = SC.A . AQuoted
+
+-- | Lift an integer.
+int :: Integer -> SExpr
+int = SC.A . AInt
+
+-- | Lift a natural
+nat :: Natural -> SExpr
+nat = SC.A . ANat
+
+-- | Lift a real
+real :: Rational -> SExpr
+real = SC.A . AReal
+
+-- | Lift a float
+float :: W4.FloatPrecisionRepr fpp -> BF.BigFloat -> SExpr
+float rep bf = SC.A (AFloat (Some rep) bf)
+
+-- | Lift a bitvector.
+bitvec :: Natural -> Integer -> SExpr
+bitvec w v = SC.A $ ABV (fromEnum w) v
+
+-- | Lift a boolean.
+bool :: Bool -> SExpr
+bool = SC.A . ABool
+
+
+-- * Output of the S-Expression Formula language
+
+
+-- | Generates the the S-expression tokens represented by the sexpr
+-- argument, preceeded by a list of strings output as comments.
+printSExpr :: Seq.Seq String -> SExpr -> T.Text
+printSExpr comments sexpr =
+  let outputFmt = SC.setIndentAmount 1 $ SC.unconstrainedPrint printAtom
+  in formatComment comments <> (SC.encodeOne outputFmt $ SC.fromWellFormed sexpr)
+
+
+formatComment :: Seq.Seq String -> T.Text
+formatComment c
+  | Seq.null c = T.empty
+  | otherwise = T.pack $ unlines $ fmap formatLine (F.toList c)
+  where
+    formatLine l = printf ";; %s" l
+
+
+printAtom :: Atom -> T.Text
+printAtom a =
+  case a of
+    AId s -> s
+    --AQuoted s -> T.pack ('\'' : s)
```

Comment:
Can this stray comment be deleted? I don't think `AQuoted` exists anymore.

Reply:
Done